### PR TITLE
Fix superflous shader setting updates

### LINKED ISSUE #2283: Bad render performance based on shader data
It appears when shaders are turned on, the game spend about 50% of the map rendering time updating the shaders constants.

It appears to update these shader constants once per material switch, and for every setting.

This can be improved drastically so the constants are only set when changed, for each of the shaders.

Also, the number of shaders could be reduced.

### PR DESCRIPTION
**Pull request by gregorycu**
Improve rendering performance by ~40%
Fixes #2283

### CODE REVIEW

**Comment by nerzhul** [comment]
Travis failed

**Comment by JakubVanek** [comment]
move this line:
```c
#include <IMaterialRendererServices.h>
```
to `shader.h`

**Comment by gregorycu** [comment]
Should I be rebasing this?

**Comment by Zeno-** [comment]
> Should I be rebasing this?

Not really. You should be changing it so that it builds.

**Comment by JakubVanek** [comment]
Save my diff to file outside repository, for example ../fix.diff
```
git apply ../fix.diff
```
then amend last commit or commit and rebase. Then force-push.

File: src/game.cpp [hunk 1]
```diff
@@ -1719,6 +1759,20 @@ bool Game::startup(bool *kill,
 
 	smgr->getParameters()->setAttribute(scene::OBJ_LOADER_IGNORE_MATERIAL_FILES, true);
 
+	memset(&runData, 0, sizeof runData);
```
Comment (Zeno-) [inline src/game.cpp]: Unfortunately GameRunData objects are not POD

File: src/game.cpp [hunk 2]
```diff
@@ -823,12 +840,23 @@ class GameGlobalShaderConstantSetter : public IShaderConstantSetter
...
+		, m_skyBgColorShaderSetting("skyBgColor")
+		, m_fogDistanceShaderSetting("fogDistance")
+		, m_animationTimerVertexShaderSetting("animationTimer")
+		, m_animationTimerPixelShaderSetting("animationTimer")
+		, m_dayNightRatio("dayNightRatio")
+		, m_eyePositionPixel("eyePosition")
+		, m_eyePositionVertex("eyePosition")
+		, m_baseTexture("baseTexture")
+		, m_normalTexture("normalTexture")
+		, m_useNormalmap("useNormalmap")
```
Comment (ShadowNinja) [inline src/game.cpp]: Commas go at the end of lines.

File: src/game.cpp [hunk 3]
```diff
@@ -803,13 +803,30 @@ class GameOnDemandSoundFetcher: public OnDemandSoundFetcher
 	}
 };
 
+// before 1.8 there isn't a "integer interface", only float
+#if (IRRLICHT_VERSION_MAJOR == 1 && IRRLICHT_VERSION_MINOR < 8)
+typedef irr::f32 samplerLayer_t;
+#else
+typedef irr::s32 SamplerLayer_t;
+#endif
```
Comment (ShadowNinja) [inline src/game.cpp]: 1. Is the capitalization difference intentional?
  2. You're redefining irrlicht types?!

Comment (gregorycu) [inline src/game.cpp]: 1. No, not intentional. I'll fix this.
2. No, I'm not. I'm conditionally defining SamplerLayer_t.

File: src/game.cpp (hunk 2 above)
Comment (gregorycu) [inline src/game.cpp]: I can fix this, but I don't see anything in the style guides? Are you sure?

File: src/game.cpp (hunk 1 above)
Comment (gregorycu) [inline src/game.cpp]: I've fixed this.

**Comment by rubenwardy** [comment]
If you move the function definitions from shader.h to shader.cpp you won't need to include in the header. I'm not sure which one will increase the compile time more. How many times is shader.h included?

**Comment by nerzhul** [comment]
travis always failed. Please fix your commits by rebasing your devel branch on master, compile it and give us a valid commit

**Comment by gregorycu** [comment]
@rubenwardy I believe the functions are used in game.cpp and shader.cpp. I don't understand why this is building on my machine. I'll fix this up but playing with the defines a little.

**Comment by gregorycu** [comment]
@nerzhul Zeno said i didn't need to rebase.

**Comment by nerzhul** [comment]
@gregorycu you commit only add a class but doesn't use it. Are you sure you don't forget commits ?

**Comment by gregorycu** [comment]
What the hell is going on here?

**Comment by gregorycu** [comment]
Something very bad happened somehow, I rebranched over to https://github.com/minetest/minetest/pull/2320/files

Thank you.

**Pull request closed without merging**